### DISCUSSION TITLE
Problem playing tracks over bluetooth.

### DISCUSSION THREAD

**Discussion by mooballs** [pschatzmann/arduino-audio-tools#1189] [Q&A] [answered]
If anyone could help I'd very grateful. This code used to work, Phil helped me with one issue (declaring the sample rate) and now it compiles ok and loads but whenever I try to play something over bluetooth it freaks out and reboots. I'm sure it's something stupid I'm doing but I can't work out what it is. I was using one of the multi-outputs to feed a VU meter.
Here's the code:

`#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

#include "AudioLibs/AudioA2DP.h"
#include "AudioConfigLocal.h"

BluetoothA2DPSink a2dp_sink;
A2DPStream in;  // had to copy a2dp in to out, then wrap out in volume and eq then use multioutput to get vol_info
MultiOutput out;
StreamCopy copier(out, in);  //copying in to out
VolumePrint out1;            //changed from vol.info
I2SStream out2;              //changed from i2s
//A2DPStream in;
VolumeStream volume(out);
LinearVolumeControl(lvc);  //this is for using the log pot

Equilizer3Bands eq(volume);  //must wrap the stream up correctly
//StreamCopy copier(eq, i2s); // copy in to equilizer
ConfigEquilizer3Bands cfg_eq;

#define I2S_BUFFER_SIZE 1024  //default is 512
#define I2S_BUFFER_COUNT 6    //default is 6

int VolLevel = 0;

void writeDataStream(const uint8_t *data, uint32_t length) {
  out1.write(data, length);
}

void read_data_stream(const uint8_t *data, uint32_t length) {
  eq.write(data, length);
}

// Arduino Setup
void setup() {

  Serial.begin(115200);

  out.add(out1);  //multi-outs
  out.add(out2);

  AudioLogger::instance().begin(Serial, AudioLogger::Warning);  //tried setting to Warning from Info to aid audio quality

  //i2s_out esp32 external DAC
  auto config = out2.defaultConfig(TX_MODE);
  config.pin_bck = 26;
  config.pin_ws = 25;
  config.pin_data = 23;
  config.sample_rate = 44100;  //changed from a2dp_sink.sample_rate();
  config.channels = 2;
  config.bits_per_sample = 16;
  config.buffer_size = I2S_BUFFER_SIZE;
  config.buffer_count = I2S_BUFFER_COUNT;
  out2.begin(config);

  cfg_eq = eq.defaultConfig();
  cfg_eq.setAudioInfo(config);  // use channels, bits_per_sample and sample_rate from kit

  cfg_eq.gain_medium = 0.5;

  eq.begin(cfg_eq);

  // set initial volume
  volume.begin(config);  // we need to provide the bits_per_sample and channels
  volume.setVolume(0.0);
  volume.setVolumeControl(lvc);  //for using log pot
  // register callback
  //a2dp_sink.set_pin_config(my_pin_config);
  a2dp_sink.set_stream_reader(writeDataStream, false);
  a2dp_sink.set_stream_reader(read_data_stream, false);

  a2dp_sink.set_mono_downmix(true);  //makes mono
  Serial.print("starting a2dp...");
  // Start Bluetooth Audio Receiver
  a2dp_sink.set_auto_reconnect(true);
  a2dp_sink.start("BT");  //bluetooth name
}

// Arduino loop
void loop() {
  copier.copy();  
}
`

**And here's the decoded stack results with the relevant line from the library in bold:**

0x40083a49: panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c line 408
0x40094ed1: esp_system_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c line 137
0x4009b25d: __assert_func at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/assert.c line 85
0x400d26d2: audio_tools::Vector ::operator[](int) at C:\Users\marky\OneDrive\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioBasic/Collections/Vector.h line 228
             **inline T &operator[](int index) {**

0x400d33c8: audio_tools::VolumeOutput::write(unsigned char const*, unsigned int) at C:\Users\marky\OneDrive\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioOutput.h line 550
             **volumes_tmp[j] = 0;**

0x400d2745: audio_tools::MultiOutput::write(unsigned char const*, unsigned int) at C:\Users\marky\OneDrive\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioIO.h line 165
             **int written = vector[j]->write(buffer + start, open);**

0x400d42d7: audio_tools::VolumeStream::write(unsigned char const*, unsigned int) at C:\Users\marky\OneDrive\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/VolumeStream.h line 149
            **return p_out->write(buffer, size);**

0x400d375b: audio_tools::Equilizer3Bands::write(unsigned char const*, unsigned int) at C:\Users\marky\OneDrive\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioFilter/Equilizer.h line 109
            **return p_print->write(data, len);**

0x400d3770: read_data_stream(unsigned char const*, unsigned int) at C:\Users\marky\OneDrive\Documents\Arduino\BTfaultfind/BTfaultfind.ino line 37
            **eq.write(data, length);**

0x401b8612: BluetoothA2DPSink::audio_data_callback(unsigned char const*, unsigned int) at C:\Users\marky\OneDrive\Documents\Arduino\libraries\ESP32-A2DP-main\src\BluetoothA2DPSink.cpp line 1173
            **(*stream_reader)(data, len);**

0x400d6167: ccall_audio_data_callback(uint8_t const*, uint32_t) at C:\Users\marky\OneDrive\Documents\Arduino\libraries\ESP32-A2DP-main\src\BluetoothA2DPSink.cpp line 1332
            **actual_bluetooth_a2dp_sink->audio_data_callback(data,len);**

0x401127dd: btc_a2dp_sink_data_ready at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/bt/host/bluedroid/btc/profile/std/a2dp/btc_a2dp_sink.c line 158
0x401096fd: osi_thread_generic_event_handler at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/bt/common/osi/thread.c line 425
0x40109693: osi_thread_run at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/bt/common/osi/thread.c line 165

**Comment by pschatzmann** [comment]
I think it is because you write to out1 w/o having it configured and started by calling begin.
But you better configure out (after calling add) which will do this automatically.

**Reply by mooballs** [reply, marked as answer]
That makes complete sense, thanks again Phil you bloody legend. I'll try it asap and report back. Just out of interest, any idea why it might have worked a while ago?
